build/chip/java/rules.gni

```diff
@@ -176,6 +176,161 @@ template("java_library") {
   }
 }
 
+# Declare a java binary target
+#
+#   sources: List of .java files included in this binary. Mutually exclusive with jar_path.
+#
+#   jar_path: A path to an existing JAR. Mutually exclusive with sources.
+#
+#   output_name: File name for the output .jar (not including extension).
+#     Defaults to the input .jar file name.
+#
+#   javac_flags: additional flags to pass to the javac compiler
+#
+template("java_binary") {
+  # Figure out the output name
+  _jar_name = target_name
+  if (defined(invoker.output_name)) {
+    _jar_name = invoker.output_name
+  } else {
+    _jar_name += ".jar"
+  }
+
+  _deps = []
+  if (defined(invoker.deps)) {
+    _deps = invoker.deps
+  }
+
+  # What files will be compiled
+  _java_files = []
+  if (defined(invoker.sources)) {
+    _java_files = invoker.sources
+  }
+
+  _is_prebuilt = defined(invoker.jar_path)
+
+  _jar_output = ""
+  _target_dir_name = get_label_info(":$target_name", "dir")
+  if (_is_prebuilt) {
+    assert(_java_files == [])
+    _jar_output = "$root_out_dir/bin/$_target_dir_name/" +
+                  get_path_info(invoker.jar_path, "name") + ".jar"
+  } else {
+    _jar_output = "$root_out_dir/bin/$_target_dir_name/$_jar_name"
+  }
+
+  # Generate a list containing the expected build_config filepath of every dependency.
+  _deps_configs = []
+  foreach(_dep, _deps) {
+    _dep_gen_dir = get_label_info(_dep, "target_gen_dir")
+    _dep_name = get_label_info(_dep, "name")
+    _dep_config = "$_dep_gen_dir/$_dep_name.json"
+    _deps_configs += [ _dep_config ]
+  }
+  _rebased_deps_configs = rebase_path(_deps_configs, root_build_dir)
+
+  # Create the name for this target's build_config.
+  _library_target_name = target_name
+  _build_config = "$target_gen_dir/$_library_target_name.json"
+  _rebased_build_config = rebase_path(_build_config, root_build_dir)
+
+  # Write the build_config file for this target.
+  _config_target_name = target_name + "_config"
+  action(_config_target_name) {
+    script = write_build_config
+
+    deps = _deps
+
+    outputs = [ _build_config ]
+    args = [
+      "--jar-path",
+      rebase_path(_jar_output, root_build_dir),
+      "--build-config",
+      _rebased_build_config,
+      "--deps-configs=$_rebased_deps_configs",
+    ]
+  }
+
+  # Building from sources - perform Java compilation and JAR creation.
+  if (!_is_prebuilt) {
+    # Additional flags
+    _javac_flags = [
+      "-Werror",
+      "-Xlint:all",
+    ]
+    if (defined(invoker.javac_flags)) {
+      _javac_flags += invoker.javac_flags
+    }
+
+    # Data deps
+    _data_deps = []
+    if (defined(invoker.data_deps)) {
+      _data_deps = invoker.data_deps
+    }
+
+    # Generates a .java file containing all sources to be compiled
```

Comment:
The `.java` comment seems a bit odd. Aren't we creating a `.sources` file that contains all sources to be compiled?

Reply:
Yes, it should be .sources file